lib/mix/lib/mix/utils.ex

```diff
@@ -32,6 +32,23 @@ defmodule Mix.Utils do
     mix_home_xdg_lookup(:user_config)
   end
 
+  @doc """
+  Gets possible location of Mix cache.
+
+  Possible locations:
+
+   * `XDG_CONFIG_HOME/mix` (if `MIX_XDG` is set)
+   * `:filename.basedir(:user_cache, "mix")`
+
+  """
+  def mix_cache do
+    if System.get_env("MIX_XDG") in ["1", "true"] do
```

Comment:
We can always use the MIX_XDG style since this is a new directory.

we should also pass the os based on os.type.

Reply:
Unless MIX_XDG is set, I think we should use OS default location, on macOS that will be ~/Library/Caches.

Couple other notes:

1. :filename.basedir/2 automatically infers the OS
2. :filename.basedir/2 will use `XDG_CACHE_HOME` but only if you happen to be on Linux (or you force linux using :filename.basedir/3)

```
~% XDG_CACHE_HOME=/cache elixir -e 'IO.inspect :filename.basedir(:user_cache, "mix")'
"/Users/wojtek/Library/Caches/mix"
~% XDG_CACHE_HOME=/cache elixir -e 'IO.inspect :filename.basedir(:user_cache, "mix", %{os: :linux})'
"/cache/mix"
```

I tried clarifying this in https://github.com/elixir-lang/elixir/pull/10825/commits/f301eaabe888ac9c6f162f197a96c60624216497.